Source/Session.swift

```diff
@@ -471,12 +471,25 @@ open class Session {
     }
 
     func resumeOrSuspendTask(_ task: URLSessionTask, ifNecessaryForRequest request: Request) {
-        if startRequestsImmediately || request.isResumed {
+        if startRequestsImmediately && request.isInitialized {
+            // Resumes the request and ensures the state is updated to .resumed
+            request.resume()
+            return
+        }
+
+        // There's no way right now to avoid calling didResume or didSuspend twice in this case.
+        // The race is created by creating a request and calling resume (or suspend) on it right afterwards.
+        // Depending on conditions, you may end up resuming or suspending the task twice. The only way to guarantee
+        // this would only resume the task once would be to add thread-safety around the `resume`, `suspend`,
+        // and `cancel` APIs on Request. That thread-safety would also have to lock the actual calls to the
+        // task as well.
+
+        if request.isResumed && task.state != .running {
```

Comment:
Resuming an already `.running` task has no negative side effects, so are we just preventing the call to `request.didResume()` here? Same for `.suspended`? In that case, perhaps we should move the `did*` calls that trigger the event monitor into `Request` itself and have separate methods and events for the task actions. Something like `didResumeTask(_ task: URLSessionTask)` should work. That way the `Request` will only receive `didResume` whenever its state is actually updated (which should only be once no matter how many times `resume()` is called I think) but will receive `didResumeTask()` every time a task is resumed on its behalf. Same with the other state transitions.

Reply:
Yep, I think that makes a lot of sense. I think the `canTransitionToState` logic also needs to move to a testAndSet block instead of getting the value, comparing it, then setting it. It's currently not thread-safe and I've already been able to easily produce contention in a simple set of tests.

Reply:
If you have tests that hit the thread sanitizer that's great. Do you think that this PR should include the lifetime reconsiderations that were mentioned in #2778? I think separating the `Request` and task lifetimes is a start, but like you said, we probably want to revisit whether `resume()`, `cancel()`, and `suspend()` are wholly async to keep them on the same internal queue, or whether there are parts we want synchronous but threadsafe. We'll probably want to make `canTransitionToState` safe in any event. I'll give it more thought.